frontend/app.py

```diff
@@ -0,0 +1,247 @@
+    start_at = int((datetime.now() - timedelta(weeks=1)).timestamp())
+    granularity = "week"
+
+    conn = http.client.HTTPSConnection("api.enphaseenergy.com")
+    headers = {
+        "Authorization": f"Bearer {str(access_token)}",
+        "key": str(api_key)
+    }
+
+    url = f"/api/v4/systems/{enphase_system_id}/telemetry/production_micro?start_at={start_at}&granularity={granularity}"
+    conn.request("GET", url, headers=headers)
+    res = conn.getresponse()
+    data = res.read()
+    decoded_data = data.decode("utf-8")
+    data_json = json.loads(decoded_data)
+
+    return process_enphase_data(data_json, start_at)
+
+def enphase_authorization():
+    if 'access_token' not in st.session_state:
+        auth_url = get_enphase_auth_url()
+        st.write("Please visit the following URL to authorize the application:")
+        st.markdown(f"[Enphase Authorization URL]({auth_url})")
+        st.write("After authorization, you will be redirected to a URL. Please copy the entire URL and paste it below:")
+
+        redirect_url = st.text_input("Enter the redirect URL:")
+
+        if redirect_url:
+            if "?code=" not in redirect_url:
+                st.error("Invalid redirect URL. Please make sure you copied the entire URL.")
+                return None, None
+
+            auth_code = redirect_url.split("?code=")[1]
+
+            try:
+                access_token = get_enphase_access_token(auth_code)
+                st.session_state['access_token'] = access_token
+                return access_token, os.getenv('ENPHASE_SYSTEM_ID')
+            except Exception as e:
+                st.error(f"Error getting access token: {str(e)}")
+                return None, None
+    else:
+        return st.session_state['access_token'], os.getenv('ENPHASE_SYSTEM_ID')
+
+    return None, None
+
+def make_pv_data(site: PVSite, ts: pd.Timestamp, access_token: str = None, enphase_system_id: str = None) -> xr.Dataset:
+    live_generation_kw = None
+
+    if site.inverter_type == 'enphase' and access_token and enphase_system_id:
+        live_generation_kw = get_enphase_data(enphase_system_id, access_token)
+
+    da = process_pv_data(live_generation_kw, ts, site)
+    return da
+
+def predict_ocf(site: PVSite, model=None, ts: datetime | str = None, nwp_source: str = "icon", access_token: str = None, enphase_system_id: str = None):
+    if ts is None:
+        ts = pd.Timestamp.now().round("15min")
+    if isinstance(ts, str):
+        ts = datetime.fromisoformat(ts)
+
+    nwp_xr = get_nwp(site=site, ts=ts, nwp_source=nwp_source)
+    pv_xr = make_pv_data(site=site, ts=ts, access_token=access_token, enphase_system_id=enphase_system_id)
+
+    pred_df = forecast_v1_tilt_orientation(nwp_source, nwp_xr, pv_xr, ts, model=model)
+    return pred_df
+
+def run_forecast(site: PVSite, model: str = "gb", ts: datetime | str = None, nwp_source: str = "icon", access_token: str = None, enphase_system_id: str = None) -> pd.DataFrame:
```

Comment:
This function can be imported from the `forecast.py` file instead of duplicated from here?

Reply:
> This function can be imported from the `forecast.py` file instead of duplicated from here?

This function uses the predict_ocf function that uses make_pv_data from data.py, which uses some CLI based functions from enphase.py, so to avoid that I modified this here